### ISSUE TITLE
Game Audio - Very Low Volume

### ISSUE THREAD

**Issue by duda117** [YoYoGames/GameMaker-Bugs#4988]
### Description
I am developing a mobile game and have noticed that the audio volume when exporting to Android devices is very low.

In this report I have attatched a .OGG music track from my game as well as an executable APK which is just an empty room with nothing but the same .OGG music track playing.

If you install the APK on an android device you will hear the music track playing, then compare that to directly playing the .OGG file using the device's audio player. You will notice the huge difference in volume, almost 8-10 times less when playing on the APK.

I have no idea why this is happening and want to know if there is something I'm doing wrong or this is a engine bug.

My project is just an empty room with 1 object playing the music using the command audio_play_sound(). For the imported music track I have tried all the compression options , as well as adjusting the sample rate , bit rate and mono/stereo output but none of these seemed to influence the very low volume on the final APK file.

At first I thought my android tablet might have low quality speakers but I've tried it on multiple devices with no success. Also the fact that the .OGG plays well directly using the audio player on device means that the problem cant be with the speaker quality.

One more IMPORTANT thing to mention is that this problem completely goes away when I use headphones on my android device. With headphones plugged in the volume and quality of the music track is the same when playing the .OGG from the music player or from the APK file. I googled the problem and came across this post from another game engine with a similar problem.

https://discuss.cocos2d-x.org/t/sound-on-native-android-is-too-low-with-many-audio-files/58875/10

They seem to have found a solution , but I am not technically savvy enough to understand what is being suggested in the post. I hope it can help and make sense to someone with more understanding of the GameMaker audio engine.

### Steps To Reproduce
1. Download the .OGG file onto an Android device and play it directly from the device.
2. Install the APK on the same Android device and run it. It will start playing the same music.
3. Notice the volume of the music while played through the APK is around 8-10 times lower through the same device speakers.
4. Plug headphones into the device and repeat the above steps , you will notice the problem is gone , both .OGG and APK have the same volume and quality

[7bcf1d9b-d700-43fd-87a9-727e4326ec0b](https://api.gamemaker.io/api/github/downloads/7bcf1d9b-d700-43fd-87a9-727e4326ec0b)

**Comment by toby-yoyo** [comment]
There are three sides to this:
1. Downmixing at compilation - the example project has a stereo sound which is downmixed to mono during compilation, and then upmixed back to stereo at runtime. The downmix is a lossy transformation and results in an overall level reduction. Setting the sound to be exported as stereo will avoid this.
2. Power preservation at runtime - when up/downmixing at runtime, we attempt to preserve the overall level of the signal. However, we did find that there is sometimes channel-based gain scaling applied in cases where no up/downmix occurs. An example of this would be mixing a stereo file to a stereo output device, for which the output comes out around 3dB quieter than the input.
3. Post processing by other apps - other applications, operating systems or audio devices may benefit from or employ some level of post-processing to audio, which can give the wrong impression of what the 'correct' level is, which from our perspective, is the level measured from the signal contained in the original audio file.

From our measurements, we are finding that in a pass-through setup, we are outputting a signal that is around 3dB quieter than what we are putting in. This is only a fraction of what would be needed to bring the level up to what is heard in other music players, so it is very likely that other larger factors are at play here (such as in point 3 above).

Regarding point 2, we are reluctant to bring about a sudden increase in output levels across the board in order not to force devs to have to rebalance their games. This is something that we would be more willing to address in the new runtime (GMRT), but have decided against acting on for the current runtime, and because of this, I will be closing this issue.

**Comment by duda117** [comment]
Hey Toby,

I have contacted my audio designer, and he says that the sound was in Mono. The audio clip in the game settings was also set to Mono, so I don't know why this was mentioned as a reason for the problem.

Anyway I tried with the same sound exported in Stereo and using Stereo audio clip settings, but that still didn't change the outcome at all.

The same .ogg file when played from device media player is almost 8-10 times louder than when played from the .apk. This is not just a small -3dB difference like you mentioned. The difference is huge and literally ruins the listening experience, especially in my game where audio is very important.

I am a bit confused by your response and quick closure of the bug. Did you confirm the sound difference between the android device media player and apk volume? Are you saying on your device you found only a 3dB difference ?

Also I will happily wait if this can be fixed in the new runtime (GMRT). It's horrible to have my games audio gimped in comparison to other similar games in the genre, a huge disadvantage by using game maker. Really hope this can be addressed.

**Comment by toby-yoyo** [comment]
Hey, I'll try to clarify each point with reference to the previous reply.

> I have contacted my audio designer, and he says that the sound was in Mono. The audio clip in the game settings was also set to Mono, so I don't know why this was mentioned as a reason for the problem.

The original file is stereo - here it is in an analyser showing that there are two channels in the stream:

![image](https://github.com/YoYoGames/GameMaker-Bugs/assets/98745055/17804ddd-aa39-4ea4-b358-c6b2434bb48e)

As mentioned in the first point of the last post, the conversion to mono (at compile time) and then back to stereo (at runtime) means this conversion can be avoided with the correct settings, and a 2dB loss (which was measured) can be avoided. The reason this was brought up was that it is within your control to resolve. You are correct that it won't make a huge difference, but these things add up, and it is part of the problem you highlighted and that we are attempting to explain.

> Did you confirm the sound difference between the android device media player and apk volume? Are you saying on your device you found only a 3dB difference ?

The stream in that original audio file comes out to -16dB for each channel. So that is what we are considering the 'true' level of the audio stream. Given no loss during compilation (by avoiding the conversion to mono), we were giving the output device a stream measured at -19dB. This 3dB loss at mix-time was mentioned in the second point in the last post as again, it makes up part of the problem we are trying to explain and it is something we want to let you know you cannot do anything about.

> This is not just a small -3dB difference like you mentioned. The difference is huge and literally ruins the listening experience, especially in my game where audio is very important.

This is down to point 3 in the last post - and it is by far the most important point in all of this. Media playback apps on more modern devices can prompt Android into post-processing audio output. So what you are considering the 'correct' level is something that has gone through a ton of post-processing which has significantly boosted the level of the stream. What you hear in a media playback app is not the data in the audio file 'as is'.

GameMaker currently does not benefit from this because the audio backend which we have been using (for backwards compatibility) does not support it. However, in the last release (2024.4) we increased the minimum supported Android SDK version from 16 to 21 and this may allow us to move to using a backend which does support this. This is quite a big change so I have raised a feature request on your behalf, which will be discussed by the team - see https://github.com/YoYoGames/GameMaker-Bugs/issues/6024.
